aptos-move/mvhashmap/src/lib.rs

```diff
@@ -105,30 +167,83 @@ impl<K: Hash + Clone + Eq, V> MVHashMap<K, V> {
-    pub fn read(&self, key: &K, txn_idx: TxnIndex) -> Result<(Version, Arc<V>), Option<TxnIndex>> {
+    /// Read entry from transaction 'txn_idx' at access path 'key'.
+    pub fn read(&self, key: &K, txn_idx: TxnIndex) -> Result<V> {
+        use MVHashMapError::*;
+        use MVHashMapOutput::*;
+
         match self.data.get(key) {
             Some(tree) => {
-                // Find the dependency
                 let mut iter = tree.range(0..txn_idx);
-                if let Some((idx, write_cell)) = iter.next_back() {
-                    let flag = write_cell.flag();
+
+                // If read encounters a delta, it must traverse the block of transactions
+                // (top-down) until it encounters a write or reaches the end of the block.
+                // During traversal, all aggregator deltas have to be accumulated together.
+                let mut accumulator: Option<DeltaOp> = None;
+                while let Some((idx, entry)) = iter.next_back() {
+                    let flag = entry.flag();
 
                     if flag == FLAG_ESTIMATE {
                         // Found a dependency.
-                        Err(Some(*idx))
-                    } else {
-                        debug_assert!(flag == FLAG_DONE);
-                        // The entry is populated, return its contents.
-                        let write_version = (*idx, write_cell.incarnation);
-                        Ok((write_version, write_cell.data.clone()))
+                        return Err(Dependency(*idx));
+                    }
+
+                    // The entry should be populated.
+                    debug_assert!(flag == FLAG_DONE);
+
+                    match (&entry.cell, accumulator.as_mut()) {
+                        (EntryCell::Write(incarnation, data), None) => {
+                            // Resolve to the write if no deltas were applied in between.
+                            let write_version = (*idx, *incarnation);
+                            return Ok(Version(write_version, data.clone()));
+                        }
+                        (EntryCell::Write(incarnation, data), Some(accumulator)) => {
+                            // Deltas were applied. We must deserialize the value
+                            // of the write and apply the aggregated delta accumulator.
+
+                            // None if data represents deletion. Otherwise, panics if the
+                            // data can't be resolved to an aggregator value.
+                            let maybe_value = AggregatorValue::from_write(data.as_ref());
+
+                            if maybe_value.is_none() {
+                                // Resolve to the write if the WriteOp was deletion
+                                // (MoveVM will observe 'deletion').
+                                let write_version = (*idx, *incarnation);
+                                return Ok(Version(write_version, data.clone()));
```

Comment:
this is what the vm would do https://github.com/aptos-labs/aptos-core/blob/634e76e8da009f7f7504764c312a94c8a047d378/aptos-move/aptos-aggregator/src/aggregator_extension.rs#L225

validation should correct it to abort on reading the aggregator resource (not the aggregator value).